### ISSUE TITLE
UHD-3.12.0.0 with boost-1.66.0 static build failed

### ISSUE THREAD

**Issue by amore-amore** [EttusResearch/uhd#187]
After patching device3_impl.cpp by adding #include <../lib/usrp/device3/device3_impl.hpp>
build stuck using GCC 4.9.4 on device3_io_impl.cpp with error below.

I can test any changes to make it working.

```
-- 
-- Configuring the python interpreter...
-- Python interpreter: /usr/local/bin/python
-- Override with: -DPYTHON_EXECUTABLE=<path-to-python>
fatal: Not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
-- Could not determine git branch. Probably building from tarball.
fatal: Not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
-- Using UHD Images Directory: OFF
-- Building Static Libraries: ON
-- Build type not specified: defaulting to release.
-- 
-- Configuring Boost C++ Libraries...
CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:801 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:907 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1542 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:326 (FIND_PACKAGE)


CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:801 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:907 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1542 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:326 (FIND_PACKAGE)


CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:801 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:907 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1542 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:326 (FIND_PACKAGE)


CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:801 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:907 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1542 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:326 (FIND_PACKAGE)


CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:801 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:907 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1542 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:326 (FIND_PACKAGE)


CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:801 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:907 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1542 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:326 (FIND_PACKAGE)


CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:801 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:907 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1542 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:326 (FIND_PACKAGE)


CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:801 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:907 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1542 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:326 (FIND_PACKAGE)


CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:801 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:907 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1542 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:326 (FIND_PACKAGE)


CMake Warning at /usr/local/share/cmake/Modules/FindBoost.cmake:801 (message):
  New Boost version may have incorrect or missing dependencies and imported
  targets
Call Stack (most recent call first):
  /usr/local/share/cmake/Modules/FindBoost.cmake:907 (_Boost_COMPONENT_DEPENDENCIES)
  /usr/local/share/cmake/Modules/FindBoost.cmake:1542 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:326 (FIND_PACKAGE)


-- Boost version: 1.66.0
-- Found the following Boost libraries:
--   chrono
--   date_time
--   filesystem
--   program_options
--   regex
--   system
--   unit_test_framework
--   serialization
--   thread
-- Boost include directories: /home/build/uhd-3.12.0.0/host/build/boost/include
-- Boost library directories: /home/build/uhd-3.12.0.0/host/build/boost/lib
-- Boost libraries: /home/build/uhd-3.12.0.0/host/build/boost/lib/libboost_chrono-mt.a;/home/build/uhd-3.12.0.0/host/build/boost/lib/libboost_date_time-mt.a;/home/build/uhd-3.12.0.0/host/build/boost/lib/libboost_filesystem-mt.a;/home/build/uhd-3.12.0.0/host/build/boost/lib/libboost_program_options-mt.a;/home/build/uhd-3.12.0.0/host/build/boost/lib/libboost_regex-mt.a;/home/build/uhd-3.12.0.0/host/build/boost/lib/libboost_system-mt.a;/home/build/uhd-3.12.0.0/host/build/boost/lib/libboost_unit_test_framework-mt.a;/home/build/uhd-3.12.0.0/host/build/boost/lib/libboost_serialization-mt.a;/home/build/uhd-3.12.0.0/host/build/boost/lib/libboost_thread-mt.a
-- 
-- Python checking for Python version 2.7 or greater
-- Python checking for Python version 2.7 or greater - found
-- 
-- Python checking for Mako templates 0.4.2 or greater
-- Python checking for Mako templates 0.4.2 or greater - found
-- 
-- Python checking for requests 2.0 or greater
-- Python checking for requests 2.0 or greater - "import requests" failed
-- 
-- Configuring LibUHD support...
--   Dependency Boost_FOUND = 1
--   Dependency HAVE_PYTHON_PLAT_MIN_VERSION = TRUE
--   Dependency HAVE_PYTHON_MODULE_MAKO = TRUE
--   Enabling LibUHD support.
--   Override with -DENABLE_LIBUHD=ON/OFF
-- 
-- Configuring LibUHD - C API support...
--   Dependency ENABLE_LIBUHD = ON
--   Enabling LibUHD - C API support.
--   Override with -DENABLE_C_API=ON/OFF
-- 
-- Configuring Examples support...
--   Dependency ENABLE_LIBUHD = ON
--   Enabling Examples support.
--   Override with -DENABLE_EXAMPLES=ON/OFF
-- 
-- Configuring Utils support...
--   Dependency ENABLE_LIBUHD = ON
--   Enabling Utils support.
--   Override with -DENABLE_UTILS=ON/OFF
-- 
-- Configuring Tests support...
--   Dependency ENABLE_LIBUHD = ON
--   Enabling Tests support.
--   Override with -DENABLE_TESTS=ON/OFF
-- 
-- Found FPGA submodule: /home/build/uhd-3.12.0.0/fpga-src
-- 
-- Checking for module 'libgps'
--   
-- Checking for module 'libgps >= 3.11'
--   
-- Could NOT find LIBGPS (missing: LIBGPS_LIBRARY LIBGPS_INCLUDE_DIR) 
-- Could NOT find LIBERIO (missing: LIBERIO_LIBRARY LIBERIO_INCLUDE_DIR) 
-- 
-- Configuring LIBERIO support...
--   Dependency ENABLE_LIBUHD = ON
--   Dependency LIBERIO_FOUND = FALSE
--   Disabling LIBERIO support.
--   Override with -DENABLE_LIBERIO=ON/OFF
-- 
-- Configuring USB support...
--   Dependency ENABLE_LIBUHD = ON
--   Dependency LIBUSB_FOUND = TRUE
--   Enabling USB support.
--   Override with -DENABLE_USB=ON/OFF
-- 
-- Configuring GPSD support...
--   Dependency ENABLE_LIBUHD = ON
--   Dependency ENABLE_GPSD = OFF
--   Dependency LIBGPS_FOUND = FALSE
--   Disabling GPSD support.
--   Override with -DENABLE_GPSD=ON/OFF
-- 
-- Configuring B100 support...
--   Dependency ENABLE_LIBUHD = ON
--   Dependency ENABLE_USB = ON
--   Disabling B100 support.
--   Override with -DENABLE_B100=ON/OFF
-- 
-- Configuring B200 support...
--   Dependency ENABLE_LIBUHD = ON
--   Dependency ENABLE_USB = ON
--   Enabling B200 support.
--   Override with -DENABLE_B200=ON/OFF
-- 
-- Configuring E300 support...
--   Dependency ENABLE_LIBUHD = ON
--   Disabling E300 support.
--   Override with -DENABLE_E300=ON/OFF
-- 
-- Configuring USRP1 support...
--   Dependency ENABLE_LIBUHD = ON
--   Dependency ENABLE_USB = ON
--   Disabling USRP1 support.
--   Override with -DENABLE_USRP1=ON/OFF
-- 
-- Configuring USRP2 support...
--   Dependency ENABLE_LIBUHD = ON
--   Enabling USRP2 support.
--   Override with -DENABLE_USRP2=ON/OFF
-- 
-- Configuring X300 support...
--   Dependency ENABLE_LIBUHD = ON
--   Disabling X300 support.
--   Override with -DENABLE_X300=ON/OFF
-- 
-- Configuring N230 support...
--   Dependency ENABLE_LIBUHD = ON
--   Enabling N230 support.
--   Override with -DENABLE_N230=ON/OFF
-- 
-- Configuring MPMD support...
--   Dependency ENABLE_LIBUHD = ON
--   Enabling MPMD support.
--   Override with -DENABLE_MPMD=ON/OFF
-- 
-- Configuring OctoClock support...
--   Dependency ENABLE_LIBUHD = ON
--   Disabling OctoClock support.
--   Override with -DENABLE_OCTOCLOCK=ON/OFF
-- 
-- 
-- Could NOT find UDev (missing: UDEV_INCLUDE_DIR UDEV_LIBS) 
-- 
-- Configuring priority scheduling...
--   Priority scheduling supported through pthread_setschedparam.
--   Setting thread names is not supported.
-- 
-- Configuring high resolution timing...
--   High resolution timing supported though microsec_clock.
-- 
-- Configuring module loading...
--   Module loading supported through dlopen.
-- 
-- USB support enabled via libusb.
-- 
-- Configuring interface address discovery...
--   Interface address discovery supported through getifaddrs.
-- 
-- Loading build info.
-- 
-- Adding B2XX device test target
-- Adding N3XX device test target
-- 
-- Could NOT find UDev (missing: UDEV_INCLUDE_DIR UDEV_LIBS) 
CMake Warning at utils/CMakeLists.txt:123 (MESSAGE):
  Python module `requests' not found -- uhd_images_downloader.py will not
  work without it.


CMake Warning at utils/CMakeLists.txt:124 (MESSAGE):
  You may be able to install this by running 'pip install requests'


-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
-- 
-- Configuring Manual support...
--   Dependency DOXYGEN_FOUND = NO
--   Disabling Manual support.
--   Override with -DENABLE_MANUAL=ON/OFF
-- 
-- Configuring API/Doxygen support...
--   Dependency DOXYGEN_FOUND = NO
--   Disabling API/Doxygen support.
--   Override with -DENABLE_DOXYGEN=ON/OFF
-- 
-- Configuring Man Pages support...
--   Dependency GZIP_FOUND = TRUE
--   Dependency NOT_WIN32 = TRUE
--   Enabling Man Pages support.
--   Override with -DENABLE_MAN_PAGES=ON/OFF
-- 
-- ######################################################
-- # UHD enabled components                              
-- ######################################################
--   * LibUHD
--   * LibUHD - C API
--   * Examples
--   * Utils
--   * Tests
--   * USB
--   * B200
--   * USRP2
--   * N230
--   * MPMD
--   * Man Pages
-- 
-- ######################################################
-- # UHD disabled components                             
-- ######################################################
--   * LIBERIO
--   * GPSD
--   * B100
--   * E300
--   * USRP1
--   * X300
--   * OctoClock
--   * Manual
--   * API/Doxygen
-- 
-- Building version: 3.12.0.0-0-unknown
-- Using install prefix: /usr/local
-- Configuring done
-- Generating done
-- Build files have been written to: /home/build/uhd-3.12.0.0/host/build
[  2%] Built target uhd_rpclib
Scanning dependencies of target uhd_static
[  2%] Building CXX object lib/CMakeFiles/uhd_static.dir/usrp/device3/device3_io_impl.cpp.o
In file included from /home/build/uhd-3.12.0.0/host/lib/usrp/device3/device3_io_impl.cpp:10:
In file included from /home/build/uhd-3.12.0.0/host/include/../lib/usrp/device3/device3_impl.hpp:15:
In file included from /home/build/uhd-3.12.0.0/host/include/uhd/transport/bounded_buffer.hpp:11:
In file included from /home/build/uhd-3.12.0.0/host/include/uhd/transport/bounded_buffer.ipp:14:
In file included from /home/build/uhd-3.12.0.0/host/build/boost/include/boost/function.hpp:70:
In file included from /home/build/uhd-3.12.0.0/host/build/boost/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:52:
In file included from /home/build/uhd-3.12.0.0/host/build/boost/include/boost/function/detail/function_iterate.hpp:14:
In file included from /home/build/uhd-3.12.0.0/host/build/boost/include/boost/function/detail/maybe_include.hpp:22:
/home/build/uhd-3.12.0.0/host/build/boost/include/boost/function/function_template.hpp:138:18: error: called object type 'nullptr_t' is not a function or function pointer
          return (*f)(BOOST_FUNCTION_ARGS);
                 ^~~~
/home/build/uhd-3.12.0.0/host/build/boost/include/boost/function/function_template.hpp:925:53: note: in instantiation of member function 'boost::detail::function::function_obj_invoker1<nullptr_t, bool, boost::intrusive_ptr<uhd::transport::managed_buffer> >::invoke' requested here
        { { &manager_type::manage }, &invoker_type::invoke };
                                                    ^
/home/build/uhd-3.12.0.0/host/build/boost/include/boost/function/function_template.hpp:716:13: note: in instantiation of function template specialization 'boost::function1<bool, boost::intrusive_ptr<uhd::transport::managed_buffer> >::assign_to<nullptr_t>' requested here
      this->assign_to(f);
            ^
/home/build/uhd-3.12.0.0/host/build/boost/include/boost/function/function_template.hpp:1061:5: note: in instantiation of function template specialization 'boost::function1<bool, boost::intrusive_ptr<uhd::transport::managed_buffer> >::function1<nullptr_t>' requested here
    base_type(f)
    ^
/home/build/uhd-3.12.0.0/host/lib/usrp/device3/device3_io_impl.cpp:892:13: note: in instantiation of function template specialization 'boost::function<bool (boost::intrusive_ptr<uhd::transport::managed_buffer>)>::function<nullptr_t>' requested here
            NULL);
            ^
/usr/include/sys/_null.h:12:14: note: expanded from macro 'NULL'
#define NULL    nullptr
                ^
1 error generated.
*** Error 1 in . (lib/CMakeFiles/uhd_static.dir/build.make:3483 'lib/CMakeFiles/uhd_static.dir/usrp/device3/device3_io_impl.cpp.o': cd /home...)
*** Error 1 in . (CMakeFiles/Makefile2:122 'lib/CMakeFiles/uhd_static.dir/all')
*** Error 1 in /home/build/uhd-3.12.0.0/host/build (Makefile:163 'all')
```

**Comment by amore-amore** [comment]
The same errors while building uhd-3.13.0.0 with boost-1.66.0
------------------------------------------------------------------------
[  2%] Built target uhd_rpclib
Scanning dependencies of target uhd_static
[  2%] Building CXX object lib/CMakeFiles/uhd_static.dir/usrp/device3/device3_io_impl.cpp.o
In file included from /home/build/uhd-3.13.0.0/host/lib/usrp/device3/device3_io_impl.cpp:10:
In file included from /home/build/uhd-3.13.0.0/host/include/../lib/usrp/device3/device3_impl.hpp:15:
In file included from /home/build/uhd-3.13.0.0/host/include/uhd/transport/bounded_buffer.hpp:11:
In file included from /home/build/uhd-3.13.0.0/host/include/uhd/transport/bounded_buffer.ipp:14:
In file included from /home/build/uhd-3.13.0.0/host/build/boost/include/boost/function.hpp:70:
In file included from /home/build/uhd-3.13.0.0/host/build/boost/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:52:
In file included from /home/build/uhd-3.13.0.0/host/build/boost/include/boost/function/detail/function_iterate.hpp:14:
In file included from /home/build/uhd-3.13.0.0/host/build/boost/include/boost/function/detail/maybe_include.hpp:22:
/home/build/uhd-3.13.0.0/host/build/boost/include/boost/function/function_template.hpp:138:18: error: called object type 'nullptr_t' is not a function or function pointer
          return (*f)(BOOST_FUNCTION_ARGS);
                 ^~~~
/home/build/uhd-3.13.0.0/host/build/boost/include/boost/function/function_template.hpp:925:53: note: in instantiation of member function 'boost::detail::function::function_obj_invoker1<nullptr_t, bool, boost::intrusive_ptr<uhd::transport::managed_buffer> >::invoke' requested here
        { { &manager_type::manage }, &invoker_type::invoke };
                                                    ^
/home/build/uhd-3.13.0.0/host/build/boost/include/boost/function/function_template.hpp:716:13: note: in instantiation of function template specialization 'boost::function1<bool, boost::intrusive_ptr<uhd::transport::managed_buffer> >::assign_to<nullptr_t>' requested here
      this->assign_to(f);
            ^
/home/build/uhd-3.13.0.0/host/build/boost/include/boost/function/function_template.hpp:1061:5: note: in instantiation of function template specialization 'boost::function1<bool, boost::intrusive_ptr<uhd::transport::managed_buffer> >::function1<nullptr_t>' requested here
    base_type(f)
    ^
/home/build/uhd-3.13.0.0/host/lib/usrp/device3/device3_io_impl.cpp:892:13: note: in instantiation of function template specialization 'boost::function<bool (boost::intrusive_ptr<uhd::transport::managed_buffer>)>::function<nullptr_t>' requested here
            NULL);
            ^
/usr/include/sys/_null.h:12:14: note: expanded from macro 'NULL'
#define NULL    nullptr
                ^
1 error generated.
*** Error 1 in . (lib/CMakeFiles/uhd_static.dir/build.make:3483 'lib/CMakeFiles/uhd_static.dir/usrp/device3/device3_io_impl.cpp.o': cd /home...)
*** Error 1 in . (CMakeFiles/Makefile2:122 'lib/CMakeFiles/uhd_static.dir/all')
*** Error 1 in /home/build/uhd-3.13.0.0/host/build (Makefile:163 'all')

**Comment by mbr0wn** [comment]
I've submitted a fix for that particular compiler error. I'll keep #223 open to see about static builds.
